### PR TITLE
Implement pre-defined reset types on the server side

### PR DESCRIPTION
**Pull request by longquanzheng**
We had client side implementation(https://github.com/uber/cadence/blob/master/tools/cli/commands.go#L1714 )
 which is in-efficient and can cause issues for batch operation 

Will remove the CLI implementation after we see everything working.

### CODE REVIEW

File: service/history/workflowResetor.go [hunk 1]
```diff
@@ -548,13 +580,11 @@ func getRespondActivityTaskFailedRequestFromActivity(ai *persistence.ActivityInf
...
 
-// TODO: @shreyassrivatsan reduce the number of return parameters from this method or return a struct
 func (w *workflowResetorImpl) replayHistoryEvents(
-	decisionFinishEventID int64,
-	requestID string,
+	request *workflow.ResetWorkflowExecutionRequest,
 	prevMutableState mutableState,
 	newRunID string,
-) (forkEventVersion, wfTimeoutSecs int64, receivedSignalsAfterReset []*workflow.HistoryEvent, continueRunID string, sBuilder stateBuilder, retError error) {
+) (replayResult replayHistoryResult, retError error) {
```
Comment (shreyassrivatsan) [inline service/history/workflowResetor.go]: dont need the `replayResult` and `retError`

File: service/history/workflowResetor.go [hunk 2]
```diff
@@ -934,3 +1005,30 @@ func FindAutoResetPoint(
 	}
 	return "", nil
 }
+
+func LookupResetPoint(resetPoints *workflow.ResetPoints, binaryChecksum string) (string, int64, error) {
```
Comment (shreyassrivatsan) [inline service/history/workflowResetor.go]: tests for `LookupResetPoint`?

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -1772,86 +1772,144 @@ func (e *historyEngineImpl) SyncActivity(
...
 	}
 	domainID := domainEntry.GetInfo().ID
 
 	request := resetRequest.ResetRequest
-	if request == nil || request.WorkflowExecution == nil || len(request.WorkflowExecution.GetRunId()) == 0 || len(request.WorkflowExecution.GetWorkflowId()) == 0 {
-		retError = &workflow.BadRequestError{
-			Message: "Require workflowId and runId.",
-		}
-		return
+	retError = ValidateResetRequest(request)
```
Comment (shreyassrivatsan) [inline service/history/historyEngine.go]: nit: `if err = ValidateResetRequest(request); retError != nil {`

Comment (shreyassrivatsan) [inline service/history/historyEngine.go]: `validateResetRequest` - should be private method.

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -1772,86 +1772,144 @@ func (e *historyEngineImpl) SyncActivity(
...
+
+		// Note: some reset types may need to load another base
+		switch request.GetResetType() {
+		case workflow.ResetTypeFirstDecisionCompleted:
+			break
+		case workflow.ResetTypeLastDecisionCompleted:
+			break
+		case workflow.ResetTypeBadBinary:
+			// refill DecisionFinishEventId from mutableState
+			runID, eventID, retError = LookupResetPoint(baseMutableState.GetExecutionInfo().AutoResetPoints, request.GetBadBinaryChecksum())
```
Comment (shreyassrivatsan) [inline service/history/historyEngine.go]: `LookupResetPoint` can be private method.

File: service/history/workflowResetor_test.go [hunk 5]
```diff
@@ -181,6 +181,103 @@ func (s *resetorSuite) TearDownTest() {
 	s.mockEventsCache.AssertExpectations(s.T())
 }
 
+func (s *resetorSuite) TestResetWorkflowExecution_ValidateResetRequest() {
+	request := &workflow.ResetWorkflowExecutionRequest{}
```
Comment (shreyassrivatsan) [inline service/history/workflowResetor_test.go]: use table driven tests along with names

```
tests := []struct{
name string
request *workflow.ResetWorkflowExecutionRequest
expected error
}{
}

for _, test := range tests {
t.Run(test.name, func(t *testing.T){
validateResetRequest()
})
}
```

File: service/history/workflowResetor_test.go [hunk 6]
```diff
@@ -181,6 +181,103 @@ func (s *resetorSuite) TearDownTest() {
...
+	s.EqualError(err, "BadRequestError{Message: ResetTypeBadBinary must be provided with BadBinaryChecksum}")
+
+	// succ case for ResetType
+	request.BadBinaryChecksum = common.StringPtr("test-bad-binary-checksum")
+	err = ValidateResetRequest(request)
+	s.Nil(err)
+}
+
+func (s *resetorSuite) TestResetWorkflowExecution_LookupResetPoint() {
+	chsum := "test-checksum"
```
Comment (shreyassrivatsan) [inline service/history/workflowResetor_test.go]: Same as above, use table driven tests with sub-tests.

Comment (longquanzheng) [reply service/history/workflowResetor_test.go]: Done

File: service/history/workflowResetor_test.go (hunk 5 above)
Comment (longquanzheng) [reply service/history/workflowResetor_test.go]: Done

**Comment by longquanzheng** [comment]
> This looks good. Are you planning to write tests for ResetWorkflowExecution the case where the second call to GetWorkflowExecution fails?

Yes I think it's better to have unit tests before landing this.

**Comment by longquanzheng** [comment]
Close it for now. I will do it with another plan(resetV2)

**Pull request closed without merging**